src/lib/import-export/import/handlers/backup-handler-wpress.ts

```diff
@@ -0,0 +1,217 @@
+ */
+function readFromBuffer( buffer: Buffer, start: number, end: number ): string {
+	const _buffer = buffer.subarray( start, end );
+	return _buffer.subarray( 0, _buffer.indexOf( 0x00 ) ).toString();
+}
+
+/**
+ * Reads the header of a .wpress file.
+ *
+ * @param {fs.promises.FileHandle} fd - The file handle to read from.
+ * @returns {Promise<Header | null>} - A promise that resolves to the header or null if the end of the file is reached.
+ */
+async function readHeader( fd: fs.promises.FileHandle ): Promise< Header | null > {
+	const headerChunk = Buffer.alloc( HEADER_SIZE );
+	await fd.read( headerChunk, 0, HEADER_SIZE );
+
+	if ( Buffer.compare( headerChunk, HEADER_CHUNK_EOF ) === 0 ) {
+		return null;
+	}
+
+	const name = readFromBuffer( headerChunk, 0, 255 );
+	const size = parseInt( readFromBuffer( headerChunk, 255, 269 ), 10 );
+	const mTime = readFromBuffer( headerChunk, 269, 281 );
+	const prefix = readFromBuffer( headerChunk, 281, HEADER_SIZE );
+
+	return {
+		name,
+		size,
+		mTime,
+		prefix,
+	};
+}
+
+/**
+ * Reads a block of data from a .wpress file and writes it to a file.
+ *
+ * @param {fs.promises.FileHandle} fd - The file handle to read from.
+ * @param {Header} header - The header of the file to read.
+ * @param {string} outputPath - The path to write the file to.
+ */
+async function readBlockToFile( fd: fs.promises.FileHandle, header: Header, outputPath: string ) {
+	const outputFilePath = path.join( outputPath, header.prefix, header.name );
+	await fse.ensureDir( path.dirname( outputFilePath ) );
+	const outputStream = fs.createWriteStream( outputFilePath );
+
+	let totalBytesToRead = header.size;
+	while ( totalBytesToRead > 0 ) {
+		let bytesToRead = CHUNK_SIZE_TO_READ;
+		if ( bytesToRead > totalBytesToRead ) {
+			bytesToRead = totalBytesToRead;
+		}
+
+		if ( bytesToRead === 0 ) {
+			break;
+		}
+
+		const buffer = Buffer.alloc( bytesToRead );
+		const data = await fd.read( buffer, 0, bytesToRead );
+		outputStream.write( buffer );
+
+		totalBytesToRead -= data.bytesRead;
+	}
+
+	outputStream.close();
+}
+
+export class BackupHandlerWpress extends EventEmitter implements BackupHandler {
+	private bytesRead: number;
+	private eof: Buffer;
+
+	constructor() {
+		super();
+		this.bytesRead = 0;
+		this.eof = Buffer.alloc( HEADER_SIZE, '\0' );
+	}
+
+	/**
+	 * Lists all files in a .wpress backup file by reading the headers sequentially.
+	 *
+	 * It opens the .wpress file, reads each header to get the file names, and stores them in an array.
+	 * The function continues reading headers until it reaches the end of the file.
+	 *
+	 * @param {BackupArchiveInfo} file - The backup archive information, including the file path.
+	 * @returns {Promise<string[]>} - A promise that resolves to an array of file names.
+	 */
+	async listFiles( file: BackupArchiveInfo ): Promise< string[] > {
+		const fileNames: string[] = [];
+
+		if ( ! fs.existsSync( file.path ) ) {
```

Comment:
Following https://github.com/Automattic/studio/pull/497#discussion_r1745319502, it would be great to avoid using synchronous FS functions to avoid blocking the main process.